File: CSES/5. Range Queries/CSES-1749_List Removals.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

struct SegNode {
    SegNode *ls, *rs; // left child, right child
    int val, lazy; // value, lazy tag
    SegNode() : ls(nullptr), rs(nullptr), val(0), lazy(0) {}
};

class SegmentTree {
public:
    SegNode *root;
    
    SegmentTree() {
        root = new SegNode();
    }

    // Update the range [l, r] with value v
    void update(SegNode *node, int left, int right, int l, int r, int v) {
        if (l <= left && right <= r) {
            node->lazy += v;
            node->val += v;
            return;
        }
        pushdown(node); // Push down lazy tags
        int mid = left + ((right - left) >> 1);
        if (l <= mid) update(node->ls, left, mid, l, r, v);
        if (r > mid) update(node->rs, mid + 1, right, l, r, v);
        pushup(node); // Push up node value
    }

    // Query the range [l, r]
    int query(SegNode *node, int left, int right, int l, int r) {
        if (l <= left && right <= r) {
            return node->val;
        }
        // Ensure all lazy tags have been pushed down
        pushdown(node);
        int mid = left + ((right - left) >> 1);
        // Calculate answer: sum of range
        int ans = 0;
        if (l <= mid) ans = query(node->ls, left, mid, l, r);
        if (r > mid) ans += query(node->rs, mid + 1, right, l, r);
        return ans;
    }

    // Push down lazy tags
    void pushdown(SegNode *node) {
        if (node->ls == nullptr) node->ls = new SegNode();
        if (node->rs == nullptr) node->rs = new SegNode();
        if (node->lazy != 0) {
            node->ls->lazy += node->lazy;
            node->rs->lazy += node->lazy;
            node->ls->val += node->lazy;
            node->rs->val += node->lazy;
            node->lazy = 0; // Clear current node's lazy tag
        }
    }

    // Push up node value
    void pushup(SegNode *node) {
        // Update method: sum of range
        node->val = node->ls->val + node->rs->val;
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, q, left, right;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    SegmentTree *st = new SegmentTree();
    for (int i = 0; i < n; i++) st->update(st->root, 0, n - 1, i, i, 1);
    for (int i = 0; i < n; i++) {
        cin >> q;
        left = 0, right = n;
        while (left <= right) {
            int mid = left + ((right - left) >> 1);
            if (st->query(st->root, 0, n - 1, 0, mid) >= q) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        cout << nums[left] << " ";
        st->update(st->root, 0, n - 1, left, left, -1);
    }
    cout << endl;
    return 0;
}

File: CSES/5. Range Queries/CSES-1749_List Removals.py
import sys
sys.setrecursionlimit(int(1e6))

class SegNode:
    def __init__(self) -> None:
        self.ls = self.rs = None # left and right child
        self.val = self.lazy = 0 # value, lazy tag

class SegmentTree:
    def __init__(self):
        self.root = SegNode()
    
    # update the range [l, r] with value v
    @staticmethod
    def update(node: SegNode, left: int, right: int, l: int, r: int, v: int) -> None:
        if l <= left and right <= r:
            node.lazy += v
            node.val += v
            return
        SegmentTree.pushdown(node) # push down lazy tags
        mid = (left + right) // 2
        if l <= mid:
            SegmentTree.update(node.ls, left, mid, l, r, v)
        if r > mid:
            SegmentTree.update(node.rs, mid + 1, right, l, r, v)
        SegmentTree.pushup(node) # push up node value
 
    # query the range [l, r]
    @staticmethod
    def query(node: SegNode, left: int, right: int, l: int, r: int) -> int:
        if l <= left and right <= r:
            return node.val
        # Ensure all lazy tags have been pushed down
        SegmentTree.pushdown(node) 
        # Calculate answer: sum of range
        ans = 0
        mid = (left + right) // 2
        if l <= mid:
            ans = SegmentTree.query(node.ls, left, mid, l, r)
        if r > mid:
            ans += SegmentTree.query(node.rs, mid + 1, right, l, r)
        return ans
    
    # push down lazy tags
    @staticmethod
    def pushdown(node: SegNode) -> None:
        if node.ls is None:
            node.ls = SegNode()
        if node.rs is None:
            node.rs = SegNode()
        if node.lazy:
            node.ls.lazy += node.lazy
            node.rs.lazy += node.lazy
            node.ls.val += node.lazy
            node.rs.val += node.lazy
            node.lazy = 0
    
    # push up node value
    @staticmethod
    def pushup(node: SegNode) -> None:
        # Update method: sum of range
        node.val = node.ls.val + node.rs.val

n = int(input())
nums = list(map(int, input().split()))
queries = list(map(int, input().split()))

st = SegmentTree()

for i, x in enumerate(nums):
    SegmentTree.update(st.root, 0, n - 1, i, i, 1)

for q in queries:
    left, right = 0, n
    while left <= right:
        mid = (left + right) // 2
        # print("check", mid, SegmentTree.query(st.root, 0, n - 1, 0, mid))
        if SegmentTree.query(st.root, 0, n - 1, 0, mid) >= q:
            right = mid - 1
        else:
            left = mid + 1
    print(nums[left], end=" ")
    SegmentTree.update(st.root, 0, n - 1, left, left, -1)
print()
